core/src/main/java/org/opensearch/sql/executor/execution/PaginatedPlan.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.executor.execution;
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.opensearch.sql.ast.tree.Paginate;
+import org.opensearch.sql.ast.tree.UnresolvedPlan;
+import org.opensearch.sql.common.response.ResponseListener;
+import org.opensearch.sql.executor.ExecutionEngine;
+import org.opensearch.sql.executor.QueryId;
+import org.opensearch.sql.executor.QueryService;
+
+/**
+ * PaginatedPlan represents a page request. Dislike a regular QueryPlan,
+ * it returns paged response to the user and cursor, which allows to query
+ * next page.
+ * {@link ContinuePaginatedPlan}
+ */
+public class PaginatedPlan extends AbstractPlan {
+  final UnresolvedPlan plan;
```

Comment:
Could you elaborate why we need entirely separate code path for pagination, specifically from AST to logical node to optimizer to TableScan and IndexScan eventually. I'm wondering is such difference in query plan structure we can avoid?

Because I thought pagination is just a execution mode. For the same query plan, it can run at once or paginated or long running at background. In pagination case, I thought the only difference is supposed to be only in `OpenSearchRequestBuilder` and `OpenSearchIndexScan`. Not sure if my intuition is right but I'm thinking we may only need `OpenSearchIndexScan` aware of current request is paginated by passing `fetch_size` or `cursor` to it in some way?

Reply:
That's on me. This design, does not affect non-paged execution path, is easier to update and extend, including to other data sources.
 
The query plan is the same. 

`Paginate`  extends `UnresolvedPlan` and it's counter-parts `LogicalPaginate`, and `PaginateOperator`. Similar to `LimitOperator`, it is a place where we can control how each page is generated. We can push down pagination for OpenSearch, and probably, JDBC. but for Prometheus, AFAIK, it will need to be done in the plugin.  

Already there are  some cases that we cannot push down and therefor cannot correctly paginate. `SELECT * FROM table ORDER BY length(field)` is an example of such query. Another is `SELECT * FROM table OFFSET N LIMIT M`

`PaginatedPlan` and `ContinuePaginatedPlan` extend `AbstractPlan` and capture the different ways paginated requests are executed. First request (`PaginatedPlan`) needs to create a scroll request, subsequent requests `ContinuePaginatedPlan` need to get the existing plan "from cache" (currently encoded in the cursor) and execute it to get the next page.

The other new classes -- `PagedRequest`, `ContinuePagedRequest`, `OpenSearchPagedIndexScan` and their support classes all relate to execution.

My first approach only updated `OpenSearchIndexScan` but `OpenSearchIndexScan` already supports scroll requests . Keeping the logic correct was starting to get unwieldy.

Reply:
Thanks for the details! I think if so, it makes sense to make it part of query plan. Following what you said, I think it's okay to add Paginate, LogicalPaginate and a pushdown rule for OpenSearch as well as some changes in Request and Response.

But what I'm seeing is query plan wrapped by Paginate node is totally different code path. Just wondering if any chance we make a query plan with Paginate operator analyzed and planned as usual. Not sure if it's that much special, I feel Paginate can be just another new operator, right?

Reply:
What different code paths are you referring to? I'll discuss a few in this comment but I'm not sure if that will cover everything you had questions about.

The root decision was to create `OpenSearchPagedIndexScan` similar to `OpenSearchIndexScan`. `OpenSearchIndexScan` accounts for limit > window_size and wraps a scroll request if that's the case. It also take query size limit into account. Neither of these scenarios apply to paged requests so I chose to create a separate scan class instead of complicating the existing scan class. Doing so avoids affecting existing behaviour and simpler change set.

This decision lead to separate scan builder being necessary, and a separate set of optimizer rules. We need a different rule during optimization to create a different scan builder. I don't believe current framework supports alternate rules  for similar patterns and I wasn't sure if all the existing rules would apply as-is to paginated queries, so I opted for a separate list of optimizer rules to be on the safe side. Prometheus has similar approach.

Another source of difference is initial paged query request and subsequent requests to get more data. Initial request needs to initiate a scroll request and needs relevant data, subsequent requests only need scrollId. 

Does this cover everything?

Reply:
Yes, I was referring to changes on fundamental API and logic in `core` module. One of them is what you said in planner and optimizer. I think separating `IndexScan` class or having single IndexScan with small `if-else` inside, it's up to OpenSearch storage. However, I don't think that's the reason to duplicate all the optimizer rules.

For this example, we already have RequestBuilder inside IndexScan, right? I don't understand why we need a separate createScanBuilder and then separate optimize API for paginate query plan?

Reply:
I added a few comments. I think we should review and rethink most changes on core interface, especially:

+ PhysicalPlan: getTotalHits(), toCursor()
+ QueryService: paginationPlanner field
+ StorageEngine: getTableScan()
+ LogicalOptimizer.paginationCreate()